File: backend/src/app/checks/staleness.py
"""Agent-staleness overlay for evaluated checks.

A push instance whose agent has gone silent still serves its last-known
sub-states from the hub cache. Those values are stale — *unknown*, not
necessarily down. This module turns that into Checkmk-honest output:

* an explicit ``agent`` service (OK while fresh, WARN once silent past its
  scaled threshold) so the host summary goes yellow the moment contact is lost;
  and
* a severity ceiling on every other check while stale: CRIT is capped to WARN,
  because a "down" verdict on stale data is a guess, not a fact.

OK stays OK (we keep the last value), so a stale host reads yellow via the one
``agent`` service rather than emitting a wall of yellow across every sub-state.

The same :func:`staleness_for` also drives the ``stale`` flag on the instance
API — staleness is one concept with one source of truth.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Protocol

from app.checks.models import CheckState, ServiceCheck
from app.poller.gate import agent_age_seconds, is_stale, stale_threshold


class _InstanceLike(Protocol):
    agent_mode: bool
    agent_last_seen: datetime | None
    push_interval_seconds: int | None


class _SettingsLike(Protocol):
    push_interval_seconds: int
    agent_stale_seconds: int


@dataclass(frozen=True)
class Staleness:
    """Resolved agent freshness for one push instance."""

    stale: bool
    age_seconds: int
    threshold: int


def staleness_for(inst: _InstanceLike, settings: _SettingsLike, now: datetime) -> Staleness | None:
    """Agent staleness for a push instance, or None when the concept doesn't apply.

    Returns None for direct/relay-poll instances and for push instances whose
    agent has never reported (nothing to be stale about yet). The threshold scales
    with the instance's effective push interval, floored at ``agent_stale_seconds``
    — matching the watchdog so the API/export agree with the offline flip.
    """
    if not inst.agent_mode or inst.agent_last_seen is None:
        return None
    threshold = stale_threshold(
        inst.push_interval_seconds,
        settings.push_interval_seconds,
        settings.agent_stale_seconds,
    )
    age = agent_age_seconds(now, inst.agent_last_seen)
    return Staleness(
        stale=is_stale(now, inst.agent_last_seen, threshold),
        age_seconds=age if age is not None else 0,
        threshold=threshold,
    )


def _agent_check(s: Staleness) -> ServiceCheck:
    if s.stale:
        return ServiceCheck(
            key="agent",
            state=int(CheckState.WARN),
            summary=(
                f"Agent silent for {s.age_seconds}s (>{s.threshold}s) — sub-state data is stale"
            ),
        )
    return ServiceCheck(
        key="agent",
        state=int(CheckState.OK),
        summary=f"Agent reporting ({s.age_seconds}s ago)",
    )


def apply_staleness(checks: list[ServiceCheck], s: Staleness | None) -> list[ServiceCheck]:
    """Prepend the ``agent`` service and, while stale, cap CRIT→WARN on the rest.

    Returns the input list unchanged (same object) when ``s`` is None — a
    direct-poll instance has no agent to be stale. Never mutates the input
    checks: capped checks are produced as copies.
    """
    if s is None:
        return checks
    agent = _agent_check(s)
    if not s.stale:
        return [agent, *checks]
    capped = [
        c
        if c.state != int(CheckState.CRIT)
        else c.model_copy(
            update={
                "state": int(CheckState.WARN),
                "summary": f"{c.summary} (stale: agent silent {s.age_seconds}s)",
            }
        )
        for c in checks
    ]
    return [agent, *capped]

File: orbit/lib/orbit/checks/staleness.ex
defmodule Orbit.Checks.Staleness do
  @moduledoc """
  Agent-staleness resolution for a push instance — port of checks/staleness.py
  + poller/gate.py. A push agent gone silent still serves its last-known
  sub-states from the hub cache; those are stale (unknown, not necessarily
  down). The threshold scales with the instance's effective push interval,
  floored at `agent_stale_seconds`, matching the watchdog so API/export agree.
  """

  alias Orbit.Instances.Instance

  @default_missed_pushes 4

  defstruct [:stale, :age_seconds, :threshold]
  @type t :: %__MODULE__{stale: boolean(), age_seconds: integer(), threshold: integer()}

  @doc """
  Staleness for a push instance, or nil when the concept doesn't apply
  (direct/relay poll, or a push agent that has never reported).

  `push_default` / `stale_floor` are the effective settings values.
  """
  @spec resolve(Instance.t(), integer(), integer(), DateTime.t()) :: t() | nil
  def resolve(%Instance{} = inst, push_default, stale_floor, now) do
    if Instance.agent_mode?(inst) and inst.agent_last_seen != nil do
      threshold = stale_threshold(inst.push_interval_seconds, push_default, stale_floor)
      age = DateTime.diff(now, inst.agent_last_seen, :second)

      %__MODULE__{
        stale: age > threshold,
        age_seconds: max(age, 0),
        threshold: threshold
      }
    end
  end

  @doc "Seconds of silence tolerated: max(floor, 4 × effective push interval)."
  @spec stale_threshold(integer() | nil, integer(), integer()) :: integer()
  def stale_threshold(push_interval, default, floor) do
    max(floor, @default_missed_pushes * effective_interval(push_interval, default))
  end

  @doc """
  The point after which an instance's cached sections stopped being evidence,
  or nil while it is reporting.

  The hub cache keeps serving the last push/poll forever (it is even rehydrated
  from `status_snapshot` at boot), so a dead box's tunnels still read
  "established". Anything that draws a TIMELINE from that cache needs to know
  where the measurements end — past this timestamp the state is last-known, not
  observed, and must be drawn as unknown rather than as up or down.

  Both transports, same rule as `stale_threshold/3` — `max(floor,
  #{@default_missed_pushes} × cycle)`: a push agent goes silent at
  `agent_last_seen`, a polled box at `last_success_at`. Leaving the poll side
  out would keep painting a dead Securepoint's tunnels green — the VPN page
  deliberately lists polled boxes too. The floor is what keeps one missed
  cycle on a fast (5–15s) poll interval from flipping a whole lane to grey.

  This is deliberately a freshness question, not a reachability one, so it can
  answer "silent" earlier than `Orbit.Instances.status_bucket/3` calls a polled
  box offline (flat 5-minute window). Data being too old to draw and a box
  being declared down are different verdicts.
  """
  @spec silent_since(Instance.t(), t() | nil, integer(), integer(), DateTime.t()) ::
          DateTime.t() | nil
  def silent_since(inst, staleness, poll_default, stale_floor, now)

  def silent_since(%Instance{} = inst, %__MODULE__{stale: true}, _poll_def, _floor, _now),
    do: inst.agent_last_seen

  def silent_since(%Instance{}, %__MODULE__{}, _poll_def, _floor, _now), do: nil

  def silent_since(%Instance{} = inst, nil, poll_default, stale_floor, now) do
    # No staleness struct: a polled box (handled here) or a push agent that has
    # never reported — the latter has no cached sections to grey out at all.
    with false <- Instance.agent_mode?(inst),
         %DateTime{} = last <- inst.last_success_at,
         tolerated = stale_threshold(inst.poll_interval_seconds, poll_default, stale_floor),
         true <- DateTime.diff(now, last) > tolerated do
      last
    else
      _ -> nil
    end
  end

  defp effective_interval(override, _default) when is_integer(override) and override > 0,
    do: override

  defp effective_interval(_override, default), do: default
end
